### PR TITLE
fix: add epsilon guard to torch.log in PoissonLoss and CategoricalCrossEntropy

### LINKED ISSUE #5099: PoissonLoss and CategoricalCrossEntropy – log(0) causes -inf / NaN in PyTorch path
## 🐛 Bug

`PoissonLoss` and `CategoricalCrossEntropy` call `torch.log()` directly on raw model outputs **without any numerical guard**. When the output tensor contains `0.0` (e.g., a fully saturated softmax class or a zero predicted count), `torch.log(0)` evaluates to `-inf`, which propagates through the loss → gradient becomes `NaN` → model weights silently corrupt.

This is inconsistent with sibling loss classes in the same file (`ShannonEntropy`, `VAE_KLDivergence`) that already apply an epsilon guard.

**Affected classes in `deepchem/models/losses.py`:**

| Class | Line | Unsafe Expression |
|---|---|---|
| `PoissonLoss._create_pytorch_loss` | 158 | `torch.log(output)` |
| `CategoricalCrossEntropy._create_pytorch_loss` | 206 | `torch.log(output)` |

## To Reproduce

1. Install DeepChem with PyTorch.
2. Run the following snippet:

```python
import torch
from deepchem.models.losses import PoissonLoss, CategoricalCrossEntropy

# Case 1: PoissonLoss with a zero predicted count
loss_fn = PoissonLoss()._create_pytorch_loss()
output = torch.tensor([0.0, 1.0, 2.0])  # 0.0 is a valid Poisson prediction
labels = torch.tensor([1.0, 1.0, 1.0])
result = loss_fn(output, labels)
print("PoissonLoss result:", result)
# → tensor(nan) or tensor(-inf)

# Case 2: CategoricalCrossEntropy with a saturated zero-probability class
loss_fn2 = CategoricalCrossEntropy()._create_pytorch_loss()
output2 = torch.tensor([[0.0, 0.5, 0.5]])  # first class probability is 0
labels2 = torch.tensor([[1.0, 0.0, 0.0]])  # but it is the true class
result2 = loss_fn2(output2, labels2)
print("CategoricalCrossEntropy result:", result2)
# → tensor([nan]) or tensor([inf])
```

3. Observe that both results are non-finite (`nan` or `inf`). Any downstream `.backward()` call will corrupt all model parameters with no error raised.

## Expected behavior

Both loss functions should return a **finite scalar/tensor** for any valid input, matching the behavior of `ShannonEntropy` and `VAE_KLDivergence` which already handle this correctly in the same file:

```python
# ShannonEntropy (line 466) — correct pattern already in codebase
torch.log(1e-20 + inputs)

# VAE_KLDivergence (line 414) — correct pattern already in codebase
torch.log(1e-20 + torch.square(logvar))
```

### Proposed fix

Clamp the argument to `torch.log` to a small positive floor using `torch.finfo`, which is dtype-aware (`float16` ≈ `6e-8`, `float32` ≈ `1.2e-7`, `float64` ≈ `2.2e-16`):

```python
# PoissonLoss._create_pytorch_loss  (line 158)
eps = torch.finfo(output.dtype).eps
return torch.mean(output - labels * torch.log(output.clamp(min=eps)))

# CategoricalCrossEntropy._create_pytorch_loss  (line 206)
eps = torch.finfo(output.dtype).eps
return -torch.sum(labels * torch.log(output.clamp(min=eps)), dim=-1)
```

A regression test should also be added to `deepchem/models/tests/test_losses.py`:

```python
import torch
from deepchem.models.losses import PoissonLoss, CategoricalCrossEntropy

def test_poisson_loss_zero_output_no_nan():
    loss_fn = PoissonLoss()._create_pytorch_loss()
    output = torch.tensor([0.0, 1.0, 2.0])
    labels = torch.tensor([1.0, 1.0, 1.0])
    result = loss_fn(output, labels)
    assert torch.isfinite(result), f"PoissonLoss returned non-finite value: {result}"

def test_categorical_crossentropy_zero_prob_no_nan():
    loss_fn = CategoricalCrossEntropy()._create_pytorch_loss()
    output = torch.tensor([[0.0, 0.5, 0.5]])
    labels = torch.tensor([[1.0, 0.0, 0.0]])
    result = loss_fn(output, labels)
    assert torch.all(torch.isfinite(result)), f"CategoricalCrossEntropy returned non-finite value: {result}"
```

## Environment

* OS:
* Python version:
* DeepChem version:
* RDKit version (optional):
* TensorFlow version (optional):
* PyTorch version:
* Any other relevant information:

## Additional context

- The bug is **completely silent** — no exception is raised. The first symptom is `loss=nan` appearing in training logs after one unlucky batch, which is often misdiagnosed as a learning rate or architecture issue.
- The **TensorFlow paths are not affected** — `_compute_tf_loss` for both classes delegates to `tf.keras.losses`, which applies its own internal clamping.
- The fix is a 2-line change per class, making this a great first-contribution PR.

### PR DESCRIPTION
**Pull request by KaiOnCode**
## Description

Fixes #5099: PoissonLoss and CategoricalCrossEntropy – log(0) causes -inf / NaN in PyTorch path

Added epsilon guard to `torch.log` calls to prevent `-inf`/`NaN` when output contains zeros.

## Changes

- `PoissonLoss._create_pytorch_loss`: Added `eps = torch.finfo(output.dtype).eps` and changed `torch.log(output)` to `torch.log(output.clamp(min=eps))`
- `CategoricalCrossEntropy._create_pytorch_loss`: Added `eps = torch.finfo(output.dtype).eps` and changed `torch.log(output)` to `torch.log(output.clamp(min=eps))`

## Rationale

This matches the pattern already used in `ShannonEntropy` and `VAE_KLDivergence` in the same file. The TensorFlow paths are not affected as they delegate to `tf.keras.losses` which applies its own internal clamping.

### CODE REVIEW

**Comment by KaiOnCode** [comment]
Closing — module has shown no human activity on community PRs in the past week (zero reviews, zero comments). Revisiting if activity picks up.
